server/protocols/protocols-imap4/src/main/java/org/apache/james/imapserver/netty/ReactiveThrottler.java

```diff
@@ -0,0 +1,91 @@
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.imapserver.netty;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.james.metrics.api.GaugeRegistry;
+import org.reactivestreams.Publisher;
+
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Sinks;
+import reactor.core.scheduler.Schedulers;
+
+public class ReactiveThrottler {
+    public static class RejectedException extends RuntimeException {
+        public RejectedException(String message) {
+            super(message);
+        }
+    }
+
+    private final int maxConcurrentRequests;
+    private final int maxQueueSize;
+    // In flight + executing
+    private final AtomicInteger concurrentRequests = new AtomicInteger(0);
+    private final Queue<Publisher<Void>> queue = new ConcurrentLinkedQueue<>();
+
+    public ReactiveThrottler(GaugeRegistry gaugeRegistry, int maxConcurrentRequests, int maxQueueSize) {
+        gaugeRegistry.register("imap.request.queue.size", () -> Math.max(concurrentRequests.get() - maxConcurrentRequests, 0));
+
+        this.maxConcurrentRequests = maxConcurrentRequests;
+        this.maxQueueSize = maxQueueSize;
+    }
+
+    public Mono<Void> throttle(Publisher<Void> task) {
+        if (maxConcurrentRequests < 0) {
+            return Mono.from(task);
+        }
+        int requestNumber = concurrentRequests.incrementAndGet();
+
+        if (requestNumber <= maxConcurrentRequests) {
+            // We have capacity for one more concurrent request
+            return Mono.from(task)
+                .doFinally(any -> onRequestDone());
+        } else if (requestNumber <= maxQueueSize + maxConcurrentRequests) {
+            // Queue the request for later
+            Sinks.One<Void> one = Sinks.one();
+            queue.add(Mono.from(task)
+                .then(Mono.fromRunnable(() -> one.emitEmpty(Sinks.EmitFailureHandler.FAIL_FAST))));
+            // Let the caller await task completion
+            return one.asMono();
+        } else {
+            concurrentRequests.decrementAndGet();
+
+            return Mono.error(new RejectedException(
+                String.format(
+                    "The IMAP server has reached its maximum capacity "
+                        + "(concurrent requests: %d, queue size: %d)",
+                    maxConcurrentRequests, maxQueueSize)));
+        }
+    }
+
+    private void onRequestDone() {
+        concurrentRequests.decrementAndGet();
+        Publisher<Void> throttled = queue.poll();
+        if (throttled != null) {
+            Mono.from(throttled)
+                .doFinally(any -> onRequestDone())
+                .subscribeOn(Schedulers.parallel())
```

Comment:
I have a concern here about the `Schedulers.parallel()`: the deeper layer of our code may contain some blocking code that is not friendly with the parallel scheduler.

Reply:
It is the responsibilities to wrap such calls and execute it on an elastic scheduler.

Blocking calls on the netty event loops are critical too and should be avoided at all cost, too.

This changes nothing...